### PR TITLE
feat: pans algorithm implementation

### PR DESCRIPTION
**Pull request by BryanPan342**
This is a very loaded PR and contains a lot of moving parts:

**Features**
* Created utility functions `kLargest` and `quickSelect` for a O(n) way of getting the 50 largest elements in a list (needed for pan's algo)
* Created utility function `getNRandomElements` to grab N random elements in an array
* Expand `objectToForm` to take in a string array
* Implement spotify function to replace top songs
* Add `song metadata` to lobby object
* Implemented Pan's Algorithm as part of the lobby object `Lobby.synthesizePlaylist()`

**Fixes**
* UserId is null results in retrieving the CL account user in `verifyIds`. Thanks @darren2hang #39 
* When a new user is created the `updateTopSongs` call will create 2 users

**Tests**
* Test for utility functions `kLargest` and `quickSelect`

**Breaking Changes**
* Currently we don't have a good way of checking if the theme is a valid theme, so I just set it as `DISSOCIATING_ON_A_HIGHWAY`

Closes: #7

### CODE REVIEW

File: lib/playlist-generation/synthesize-playlist.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ *
+ * @param lobbyId the lobby to analyze
+ */
+export async function synthesizePlaylist(lobbyId: string, theme: THEME): Promise<boolean> {
+  const lobby = await Lobby.fromId(lobbyId);
+  if (!lobby) return false;
+
+  const songsMap = await lobby.participants.reduce(async (accP: Promise<Record<string, number>>, userId) => {
```
Comment (trevorong) [inline lib/playlist-generation/synthesize-playlist.ts:15]: this gets the frequencies of songs among all users in the lobby?

File: lib/playlist-generation/synthesize-playlist.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const user = await User.fromId(userId);
+    if (!user) return acc;
+
+    user.topSongs.forEach((songId) => {
+      if (!acc[songId]) acc[songId] = 0;
+      acc[songId]++;
+    });
+
+    return acc;
+  }, Promise.resolve({}));
```
Comment (trevorong) [inline lib/playlist-generation/synthesize-playlist.ts:27]: this gets the frequencies of songs among all users in the lobby?

File: lib/playlist-generation/synthesize-playlist.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    const score = computeScore(song.audioFeatures, theme) * (1 + overlapScore * .1);
+    if (score === 0) return;
+    songScores.push({id, score});
+  });
+
+  const topSongs = kLargest<Song2Score>(songScores, compareSongScores, 50);
+
+  const additionalSongs = topSongs.length === 50
+    ? []
+    : getNRandomElements(Object.keys(songsMap), 50 - topSongs.length);
```
Comment (trevorong) [inline lib/playlist-generation/synthesize-playlist.ts:43]: will this potentially grab duplicate songs? i think spotify lets u add the same song twice to a playlist

File: lib/playlist-generation/synthesize-playlist.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+const compareSongScores = (a: Song2Score, b: Song2Score) => a.score < b.score ? -1 : 1;
+
+type AudioFeatureEntry = [feature: keyof AudioFeatures, score: number];
+
+function computeScore(af: AudioFeatures, theme: THEME): number {
+  const conditions = theme2Conditions[theme];
+  const isQualifying = !Object.entries(af).find(([feature, score]: AudioFeatureEntry) => {
+    return score < conditions[feature].min || conditions[feature].max < score;
+  });
```
Comment (trevorong) [inline lib/playlist-generation/synthesize-playlist.ts:58]: does this only accept if a song meets all of the conditions for the theme?

File: lib/playlist-generation/themes/index.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import { dissociating } from './dissociating';
+
+export type ThemeConditions = {
+  min: number,
+  max: number,
+  target: number,
+  weight: number,
+}
+
+export const NO_WEIGHT: ThemeConditions = { min: 0, max: 0, target: 0, weight: 0 };
```
Comment (trevorong) [inline lib/playlist-generation/themes/index.ts:11]: won't NO_WEIGHT reject all songs because the score won't be > and < 0?  based on computeScore

Comment (BryanPan342) [reply lib/playlist-generation/themes/index.ts:11]: thats the whole point of it right?

Comment (BryanPan342) [reply lib/playlist-generation/themes/index.ts:11]: o wait no i see what ur saying nice catch

File: lib/playlist-generation/synthesize-playlist.ts (hunk 4 above)
Comment (BryanPan342) [reply lib/playlist-generation/synthesize-playlist.ts:58]: isQualifying is set to false if there exists an audio feature with a score less than min condition or greater than max condition

File: lib/playlist-generation/synthesize-playlist.ts (hunk 3 above)
Comment (BryanPan342) [reply lib/playlist-generation/synthesize-playlist.ts:43]: hm yes.. maybe thats a problem for another day HAHA

File: lib/playlist-generation/synthesize-playlist.ts (hunk 2 above)
Comment (BryanPan342) [reply lib/playlist-generation/synthesize-playlist.ts:27]: yup! it's a frequency map

File: www/lobby/index.ts [hunk 6]
```diff
@@ -16,21 +18,28 @@ lobby_router.use('/', lobby_id_router);
...
-  if(!lobby) return res.status(500).json('unable to create lobby at this time').end();
+  if(!lobby) {
+    logger.error(`Failed to create playlist, ${lobbyName}.`);
+    return res.status(500).json('unable to create lobby at this time').end();
+  }
+
+  const playlistCreated = lobby.synthesizePlaylist();
 
-  void lobby.writeToDatabase();
+  !!playlistCreated && logger.error(`Failed to synthesize playlist, ${lobbyName}.`);
```
Comment (darren2hang) [inline www/lobby/index.ts:42]: Should this be `!playlistCreated && logger.error(...` instead of `!!playlistCreated && `

**Review by darren2hang — COMMENTED** [review]
Asides from my one comment, everything looks good. I tested that the songIds are written to the database and that it returns the song list and their metadata to the GET lobby endpoint. I didn't test if the actual algorithm works

File: lib/playlist-generation/compute-score.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    return af[feature] < condition.min || condition.max < af[feature];
+  });
+
+  if (!isQualifying) return 0;
+
+  const rawScore = Object.entries(conditions).reduce((acc, [feature, {target, weight}]: AudioFeatureEntry) => {
+    return acc + (1 - Math.abs(target - af[feature])) * weight;
+  }, 0);
+
+  const numWeighted = Object.values(conditions).reduce((acc, {weight}) => acc + +(!!weight), 0);
```
Comment (darren2hang) [inline lib/playlist-generation/compute-score.ts:25]: is the `acc + +(!!weight)` a typo? Should it just be one plus sign?

File: utils/k-largest.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { quickSelect } from '.';
+import { Comparator } from './quick-select';
+
+export function kLargest<T>(arr: T[], cmp: Comparator<T>, k: number): T[] {
+  if (k > arr.length) return arr;
```
Comment (trevorong) [inline utils/k-largest.ts:5]:
```suggestion
  if (k >= arr.length) return arr;
```

File: lib/playlist-generation/compute-score.ts (hunk 7 above)
Comment (BryanPan342) [reply lib/playlist-generation/compute-score.ts:25]: the second plus is to convert the boolean to a number. Basically I want to count all the weighted features. So the first `!` allows me to turn the weight into a boolean of either 0 or 1 depending on whether weight is not 0 or is. And then the second `!` keeps it a boolean and reverts it.

**Review by darren2hang — APPROVED** [review]
LGTM!

**Review by trevorong — APPROVED** [review]
LGTM

**Pull request merged**